crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -98,11 +98,84 @@ pub(crate) struct Store {
     inner: Arc<DynCryptoStore>,
     verification_machine: VerificationMachine,
     tracked_users_cache: Arc<DashSet<OwnedUserId>>,
-    users_for_key_query_cache: Arc<DashSet<OwnedUserId>>,
+    users_for_key_query: Arc<Mutex<UsersForKeyQuery>>,
     tracked_user_loading_lock: Arc<Mutex<()>>,
     tracked_users_loaded: Arc<AtomicBool>,
 }
 
+/// Record of the users that are waiting for a /keys/query.
+///
+/// To avoid races, we maintain a sequence number which is updated each time we
+/// receive an invalidation notification. We also record the sequence number at
+/// which each user was last invalidated. Then, we attach the current sequence
+/// number to each `/keys/query` request, and when we get the response we can
+/// tell if any users have been invalidated more recently than that request.
+#[derive(Debug)]
+struct UsersForKeyQuery {
+    /// the sequence number we will assign to the next addition to user_map
+    next_seqno: InvalidationSequenceNumber,
+
+    /// the users pending a lookup, together with the sequence number at which
+    /// they were added to the list
+    user_map: HashMap<OwnedUserId, InvalidationSequenceNumber>,
+}
+
+// we use a signed int for this, so that we can do a `wrapping_sub` and get a
+// meaningful result (which is ok even if it wraps)
+type InvalidationSequenceNumber = i64;
+
+impl UsersForKeyQuery {
+    /// create a new, empty, `UsersForKeyQueryCache`
+    fn new() -> Self {
+        UsersForKeyQuery { next_seqno: 0, user_map: HashMap::new() }
+    }
+
+    /// record a new user that requires a key query
+    fn insert_user(&mut self, user: &UserId) {
+        let seq = self.next_seqno;
+        trace!(?user, sequence_number = seq, "Flagging user for key query");
+        self.user_map.insert(user.to_owned(), seq);
+        self.next_seqno += 1;
```

Comment:
This panics on debug builds and we never get to utilize the wrapping subtractions.